### PR TITLE
Sage on nixos

### PR DESCRIPTION
**Pull request by timokau**
###### Motivation for this change

The current sage package rebuilds all its dependencies with its own build system. Thats a mess:

- It takes a long time
- Its basically 50% luck it works at all. Probably lots of failing tests too, I never tried.
- Its hard to patch
- It has to re-invent some nixpkgs features to patch the build system
- Its hard to integrate with other sofware (own python etc.)
- Its big

So this PR builds all the dependencies with nix instead. All the tests are passing.

It depends on (and includes) #39555, ~#39405~, ~#38794~ (without tests), ~#38787~ and the ~mpfr update in staging~. #39405 is responsible for most of the diff.

~It is not quite ready to merge yet, because it updates to sage 8.2.rc4 (last rc) which is not released yet and I still need to fix some FIXMEs, do some TODOs, document patches and remove redundant dependencies.~

I'm creating this WIP PR because 8.2 is about to be released and I don't want anybody else wasting time updating sage.

Testing this will take a lot of rebuilding because of the mpfr update (stdenv) and the python patch. ~**If you want to review, your time is probably best spend waiting until I'm finished**~. If you want to give feedback already, that is of course welcome.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by 7c6f434c** [comment]
1. Impressive.

2. I think among the things to do before removing [WIP] is making sure you don't commit any clearly temporary files.

**Comment by timokau** [comment]
Thanks :)

Oh right the `.orig` files. I put the PR up in a bit of a hurry and missed those. Removed.

**Comment by timokau** [comment]
I'm not sure about the ofBorg error, it appears to have something to do with the node packages. I'm not sure about those anyways, I'm still hoping for some feedback from someone who knows more about node packaging.

**Review by FRidh — CHANGES_REQUESTED** [review]
Thanks @timokau , it is good to see this all together. 

I went through it, and then the changes look good to me, although I am still of the opinion we should not be adding old and unmaintained packages to Nixpkgs. Therefore, if these dependencies are really needed, I suggest you move them under the scope of `sage` and out of `python-packages.nix` and `python-modules/`.

**Comment by 7c6f434c** [comment]
@FRidh I am not sure if it is fully correct to call such packages unmaintained, as the Sage team is clearly capable and quite likely to perform maintenance on these packages if unavoidable (of course they can also drop the dependency, in which case we would find out the packages were indeed unmaintained along the way…)

On the other hand, maybe they are not worth putting in the global Python namespace, just as `liborcus` for LibreOffice lived together with it because nothing else was likely to ever use it.

**Comment by timokau** [comment]
Can somebody (@7c6f434c, @FRidh?) help me figure out how to avoid all the `inherit xyz` in `default.nix`? I think I need some of `newScope`, `makeScope`, `import pkgs.path { overlays = [ (self: super: {...})] }` and `python.pkgs.override`. But whatever I try, it either eats all my RAM in endless recursion, doesn't recognize the overrides or doesn't allow me to define new packages.

Could someone give me a minimal example how to create a new scope that overrides one python package and one regular pacakge? Something like:

```
pkgs = nixpkgs.giveMeNewScope (self: super: {
    glpk = super.glpk.overrideDerivation (...); # regular package
    numpy = super.python2.pkgs.numpy.override {}; # python package
    fantasy-py = self.python2.callPackage ./fantasy-py.nix {}; # new python package
});
```

**Comment by 7c6f434c** [comment]
I am not sure about proper structure of overrides for Nixpkgs nowadays.

Reimporting nixpkgs replacing user overlays with a sage-related one may
also be counterintuitive.

I am pretty sure you want something like:

self: super: rec {
  pkg1 = super.pkg1.override {…};

  callPackage = super.newScope self;

  pkg2 = callPackage … { };
}

**Comment by timokau** [comment]
That defines a function right? But where do I call that function / give it as an argument?

**Comment by 7c6f434c** [comment]
I am not sure — reimporting with overlays sound dirty, packageOverrides are deprecated…

**Comment by timokau** [comment]
Yeah reimporting with overlays is what the manual suggested, but I think thats probably a different use-case. Do you know who to ask?

**Comment by 7c6f434c** [comment]
Oh? Maybe look up in git who added that advice to ask if it applies in the case you want…

**Comment by timokau** [comment]
That would be @grahamc in `doc/functions.xml` and @nbp in `doc/language-frameworks/python.section.md`. Any help would be much appreciated.

**Comment by timokau** [comment]
Alright I think this is ready for review now. The only remaining blocker for merge is #39555. Once thats merged and through staging I will rebase this PR.

If I can clarify anything, let me know.

**Comment by teto** [comment]
thanks that's an awesome contribution. I wanted to use sage at some point, had some problems, wanted to fix it but gave up when facing the huge task (plus sage developers don't seem too receptive about packaging concerns). I am no maintainer but I believe merging the new packages in one or several PRs would make things easier. I've seen that a maintainer didn't want merging one (old) python package but I would advocate that sage is worth it ?

**Comment by timokau** [comment]
> thanks that's an awesome contribution.

You're welcome :)

> I wanted to use sage at some point, had some problems, wanted to fix it but gave up when facing the huge task (plus sage developers don't seem too receptive about packaging concerns).

There is *some* work ongoing upstream, but its not as much of an priority as I would like, true. I have also suggested all my relevant changes upstream[1] and at least none of them have been flat out rejected yet. So hopefully it will get better in the future.

> I am no maintainer but I believe merging the new packages in one or several PRs would make things easier. I've seen that a maintainer didn't want merging one (old) python package but I would advocate that sage is worth it ?

I have already split most new packages, package fixes and package updates in seperate PRs (lots of them). They've all been merged by now. The only things remaining are the pythong packages @FRidh doesn't want in the main tree, packages that don't make any sense without sage (like pybrial) and sage itself.

I think more or less the only thing blocking this PR is #39555, which is itself only blocked on the staging branch being a mess right now (not sure if thats still the case).

So once staging is cleaned up, #39555 can be merged into staging. Then we have to wait for staging to be merged into master and then this can be merged.


[1] https://trac.sagemath.org/query?status=closed&status=needs_info&status=needs_review&status=needs_work&status=new&status=positive_review&reporter=gh-timokau&order=priority

**Comment by timokau** [comment]
Finally all the dependencies are in master. I've cleaned up the history and rebased on nixpkgs-unstable for easier testing.

I think this is ready for test/merge.

@GrahamcOfBorg build sage.lib

**Comment by 7c6f434c** [comment]
https://git.sagemath.org/sage.git/patch?id=d6ad2e2ccd7bce3c46cd2ae2f349941808d4d825 seems to be 404

File: pkgs/applications/science/math/sage/default.nix
Comment (FRidh) [inline pkgs/applications/science/math/sage/default.nix]: no `python.pkgs` but `self`.

File: pkgs/applications/science/math/sage/default.nix
Comment (FRidh) [inline pkgs/applications/science/math/sage/default.nix]: `overrideAttrs`

File: pkgs/applications/science/math/sage/default.nix
Comment (FRidh) [inline pkgs/applications/science/math/sage/default.nix]: `overrideAttrs`

File: pkgs/applications/science/math/sage/pybrial.nix
Comment (FRidh) [inline pkgs/applications/science/math/sage/pybrial.nix]: 2 space indent

File: pkgs/applications/science/math/sage/pybrial.nix
Comment (FRidh) [inline pkgs/applications/science/math/sage/pybrial.nix]: is `sourceRoot` an option here?

File: pkgs/applications/science/math/sage/sage-env.nix
Comment (FRidh) [inline pkgs/applications/science/math/sage/sage-env.nix]: I don't follow this. Adding just `python3` is not sufficient?

File: pkgs/applications/science/math/sage/sage-with-env.nix
Comment (FRidh) [inline pkgs/applications/science/math/sage/sage-with-env.nix]: Most Python packages nowadays have a `pname` attribute. The plan is to make this mandatory.

File: pkgs/applications/science/math/sage/sage-with-env.nix
Comment (FRidh) [inline pkgs/applications/science/math/sage/sage-with-env.nix]: See also `python.pkgs.requiredPythonModules` although I don't think it is exactly what you are looking for.

File: pkgs/applications/science/math/sage/sagelib.nix
Comment (FRidh) [inline pkgs/applications/science/math/sage/sagelib.nix]: `pname = "sagelib";` instead

**Review by FRidh — COMMENTED** [review]
That's an impressive work! 

I've left some comments. As it doesn't touch any of the existing Python package sets I have no objections.

On a sidenote, considering you're using Python here a lot, and in various ways (packages, envs, ...) I am interested in knowing how you experienced the tooling. Any issues? Thinks you like to see different? Perhaps better to discuss is in another thread though.

(FYI, I'll be away on holidays and won't be able to give further feedback)

File: pkgs/applications/science/math/sage/sage-env.nix
Comment (timokau) [reply pkgs/applications/science/math/sage/sage-env.nix]: No, because then `wrapper.nix` won't be used and `PYTHONHOME` won't be re-set, see
https://github.com/NixOS/nixpkgs/blob/57bccb3cb8f6b5dea5c315fcaabdb992498b9314/pkgs/development/interpreters/python/wrapper.nix#L35

If I recall correctly that is important because sage calls the python3 binary from within python2. That means the python2 `PYTHONHOME` will be set and confuse python3.

The comment isn't very helpful though, I'll add more clarification.

File: pkgs/applications/science/math/sage/sage-with-env.nix
Comment (timokau) [reply pkgs/applications/science/math/sage/sage-with-env.nix]: I don't 100% understand what that function does. But since I need it to work for *all* packages (non-python), unfortunately I don't think I can use that.

File: pkgs/applications/science/math/sage/pybrial.nix
Comment (timokau) [reply pkgs/applications/science/math/sage/pybrial.nix]: As a small rant: I don't like that you have to explicitly specify the name the "real" source root will have after unpacking when using `sourceRoot`. For example in this case I have to specify `sourceRoot = "source/sage-brial"`. I'd prefer if the logic that is used to determine the `sourceRoot` would always be executed and just rename the `sourceRoot` to something standard, like `src`. I think ArchLinux does it that way. Than it can just be optional weather or not to additionally `cd` into `src`.

But yes, `sourceRoot` is the better option here. I don't remember why I didn't use it, if I even had a conscious reason.

**Comment by timokau** [comment]
> https://git.sagemath.org/sage.git/patch?id=d6ad2e2ccd7bce3c46cd2ae2f349941808d4d825 seems to be 404

I've fixed that link. I'm not entirely sure what caused that, but I've seen similar things happen before with the sage git server. I think it garbage-collects PR/ticket branches on force push or something.

Can we rely on hydras caching here or should I add all those patches to the `patches` folder? My goal is to ideally not have any patches in the `nixPatches` category for the next release (still in the process of upstreaming). But package-update backports will always be necessary.

@GrahamcOfBorg build sage.lib

**Comment by timokau** [comment]
@FRidh thank you for your review, I know it is asking a lot to review such a big PR. I agree with all your comments and fixed the issues.

> On a sidenote, considering you're using Python here a lot, and in various ways (packages, envs, ...) I am interested in knowing how you experienced the tooling. Any issues? Thinks you like to see different? Perhaps better to discuss is in another thread though.

Some things took a while to figure out (especially the problem that makes the empty env necessary as described above). Overall I'm pretty happy though and much prefer the "hand-curated" approach to the completely auto-generated package sets.

In the beginning I used to just set `PYTHONPATH` manually, but the performance of that is atrocious (adding around a full second to startup time). So I'm glad that python envs exist.

Minor nitpicks/questsions:

- Its a bit irritating that some conventions are a tiny bit different than the rest of nixpkgs. Close enough to confuse, but not the same. `pname`, just using `python.pkgs` instead of passing everything explicitly as arguments and `overridePythonAttrs` come to mind. I kind of prefer the way things are done in python, but it does hurt consistency a bit.

- I don't quite understand why you do batch updates. Is there any practical reason for it? Why not handle python package updates like updates for any other package? For example the recent batch broke a few things for me and it is a bit hard to find the cause when everything is batched. Of course nobody can reasonably expect you to test every single update. So why not let the regular update bot create the updates and rely on the community to test and merge them?

- I don't understand why pypi (with `fetchFromPypi`) seems to be the preferred source for packages. Wouldn't it be preferrable to pull things directly from source control if available? That cuts out the middle-man and makes it easier to quickly test with an unreleased version.

All in all I'm very happy with it, thank you for maintaining :)

**Comment by timokau** [comment]
@GrahamcOfBorg build sage.lib

Note that I'm still running the tests for the latest commit locally. Since the changes were pretty minor, I'm confident enough to push it already.

**Comment by timokau** [comment]
Seems like there is a timeout for the cython build. That's a shame, I'd love to see if ofBorg can build `sage.lib` (basically sage but no tests or wrappers).

**Comment by timokau** [comment]
Does anybody know if the ofBorg evla failure is meaningful?

**Comment by dotlambda** [comment]
@timokau Yes, the eval failure is because you removed `pkgs/development/python-modules/python-openid/default.nix` in ddd67df71d4e3304b094397cc8c1afbd13202d72. See https://github.com/NixOS/nixpkgs/commit/2e41891e6d595b9acf484d7d349b17a7b35b6923.

**Comment by timokau** [comment]
Oh that must've been a mistake while fixing some merge conflict because I assumed that was my version of python-openid. I've rebased on latest master. Since python-openid for python2 is removed there, the eval should succeed now.

**Comment by 7c6f434c** [comment]
@timokau re: batch updates: when upstream of the language ecosystem has a package manager, it makes sense to track it via batch updates because the outcome is easier to reason about when comparing to the upstream situation.

**Comment by timokau** [comment]
Why is that? Python packages can depend on specific version (ranges) and if a update builds in nixpkgs and doesn't break any reverse-dependencies, I see no reason not to include it right then.

**Comment by dotlambda** [comment]
It's simply that just a small percentage of the python packages would be updated if we didn't have batch upgrades. And since the batch upgrades only happen on the unstable channel, people encountering broken builds should expect that this happens from time to time. Before a new release, we try to stabilize the python package set.
Btw, @FRidh proposed https://groups.google.com/forum/#!topic/nix-devel/koLvsCVjQcA, which would allow maintainers to opt out of batch upgrades.

**Comment by 7c6f434c** [comment]
And a semi-related note: I think you should ask to get commit rights; while it might make sense to keep a Sage update in a PR while it is a WIP, and to discuss with @FRidh if you want a non-default version of a python package from a maintained branch, it is quite likely that reviewing your sage changes is useless because you have learnt about Sage much more than everyone else counted together. Also, I do think you have enough experience that your opinion about a PR «ah, it's just an uncontroversial update and it is ready to merge» is as good as anyone's.

**Comment by teto** [comment]
Congrats ! thanks for the awesome work to all involved.

**Comment by 7c6f434c** [comment]
@rbvermaa @domenkozar I think it is a good idea to give @timokau write access.

**Comment by 7c6f434c** [comment]
@timokau in the meantime, do you have 2FA enabled on GitHub?

**Comment by timokau** [comment]
Yes I have.

**Comment by timokau** [comment]
@rbvermaa @domenkozar was this ever discussed?

**Pull request merged**
